clouddriver-core/src/main/java/com/netflix/spinnaker/clouddriver/security/AccountDefinitionSecretManager.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2022 Apple, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.security;
+
+import com.netflix.spinnaker.kork.secrets.EncryptedSecret;
+import com.netflix.spinnaker.kork.secrets.SecretManager;
+import javax.annotation.Nonnull;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+@RequiredArgsConstructor
+public class AccountDefinitionSecretManager {
+  private final SecretManager secretManager;
+  private final AccountDefinitionAuthorizer authorizer;
+
+  public String getEncryptedSecret(String encryptedUri) {
+    return EncryptedSecret.isEncryptedFile(encryptedUri)
+        ? secretManager.decryptAsFile(encryptedUri).toString()
+        : secretManager.decrypt(encryptedUri);
+  }
+
+  public boolean canAccessAccountWithSecrets(@Nonnull String accountName) {
+    var username = SecurityContextHolder.getContext().getAuthentication().getName();
+    if (authorizer.isAdmin(username)) {
+      return true;
+    }
+    var userRoles = authorizer.getRoles(username);
+    if (userRoles.isEmpty()) {
+      return false;
+    }
+    // TODO(jvz): update with https://github.com/spinnaker/kork/pull/942
```

Comment:
Is the intent here to check whether the account contains secrets in its definition, and if so whether the user can access the given secret?

If so what would happen if say a K8s account was defined with the Kubeconfig file as a secret but the user trying to deploy to that account didn't have access to the Kubeconfig secret? Does it matter they don't have access to the secret itself if they're just deploying?

Reply:
The idea here would be that if an account is using a user secret (as being proposed in the Kork PR), then the user using the account must also be allowed to read the user secret (i.e., the user will have to be in one of the listed roles for the secret in addition to having write access to the account itself).

Reply:
Ok, not sure if I'm overthinking this but is there potential for secrets that should be restricted that would prevent somebody from deploying?

Thinking along the lines of a Spinnaker operator using the new secrets API to onboard a cluster role Kubeconfig, and then creating accounts using it. Without READ permission on that secret the accounts couldn't be deployed to, but presumably the operator wouldn't want users having read either?

Perhaps I'm misunderstanding the use case behind this!

Reply:
If a user isn't in one of the listed roles in the user secret, then they won't be able to use the account even if they have write access to the account. That's why I have this abstraction here despite it not doing much yet as I need to do two authZ checks: when storing the account and when using the account.

Reply:
Yeah I get that - my question is around whether that should be the intention. Is there a use case where a user wouldn't have permissions on a secret but would still want/need to deploy to an account?

It feels a bit weird to have to get permissions for two different entities when the end user (read: joe bloggs deploying their app) might not know anything about the secret configured on an account, just what it's called - not to mention it's extra load on Fiat which already struggles at scale.

Again I'm not sure if I'm just overthinking this or if what I'm asking makes sense 👍

Reply:
My issue with that is deploying with an account means you have arbitrary access to the secret (i.e., once you have it, there's always a way to sneak it out from there), so in an externalized user secrets model, that should not be allowed. Maybe there can be a relaxed mode (though that sort of already exists by just configuring `EncryptedSecret` URIs instead of the new `UserSecretReference` URIs).

Reply:
Yup that makes sense - falling back to EncryptedSecret.

Cheers!